zigpy/device.py

```diff
@@ -47,117 +54,150 @@ class Device(zigpy.util.LocalLogMixin, zigpy.util.ListenableMixin):
         self.neighbors = zigpy.neighbor.Neighbors(self)
-        self._node_handle = None
         self._pending = zigpy.util.Requests()
         self._relays = None
         self._skip_configuration = False
 
+    def schedule_group_membership_scan(self) -> asyncio.Task:
+        """Rescan device group's membership."""
+        if self._group_scan_task and not self._group_scan_task.done():
+            self.debug("Cancelling old group rescan")
+            self._group_scan_task.cancel()
+
+        self._group_scan_task = asyncio.create_task(self.group_membership_scan())
+        return self._group_scan_task
+
+    async def group_membership_scan(self) -> None:
+        """Sync up group membership."""
+        for ep_id, ep in self.endpoints.items():
+            if ep_id:
+                await ep.group_membership_scan()
+
     @property
     def initializing(self) -> bool:
         """Return True if device is being initialized."""
-        return bool(self._init_handle and not self._init_handle.done())
-
-    def schedule_initialize(self):
-        self.cancel_initialization()
-        self._init_handle = asyncio.create_task(self._initialize())
+        return self._initialize_task is not None and not self._initialize_task.done()
 
     def cancel_initialization(self) -> None:
         """Cancel initialization call."""
         if self.initializing:
-            LOGGER.debug("Canceling old initialize call")
-            self._init_handle.cancel()
+            self.debug("Canceling old initialize call")
+            self._initialize_task.cancel()
 
-    def schedule_group_membership_scan(self) -> None:
-        """Rescan device group's membership."""
-        if self._group_scan_handle and not self._group_scan_handle.done():
-            self.debug("Cancelling old group rescan")
-            self._group_scan_handle.cancel()
-        self._group_scan_handle = asyncio.ensure_future(self.group_membership_scan())
+    def schedule_initialize(self) -> Optional[asyncio.Task]:
+        # Already-initialized devices don't need to be re-initialized
+        if not self.is_partially_initialized:
+            self.debug("Skipping initialization, device is fully initialized")
+            self._application.device_initialized(self)
+            return None
 
-    async def group_membership_scan(self) -> None:
-        """Sync up group membership."""
-        for ep_id, ep in self.endpoints.items():
-            if ep_id:
-                await ep.group_membership_scan()
+        self.debug("Scheduling initialization")
+
+        self.cancel_initialization()
+        self._initialize_task = asyncio.create_task(self._initialize())
+
+        return self._initialize_task
+
+    @property
+    def is_partially_initialized(self):
+        """
+        Some devices were not properly initialized and need to have more information
+        requested after they have been joined.
+        """
+
+        if self.node_desc is None or not self.node_desc.is_valid:
```

Comment:
is there a case when `self.node_desc`  is None? The device class is initiallie with a NodeDescriptor object, but it is not valid and that's why it has the property `is_valid`. I guess we can remove that property, but then in order to distinguish if the node-descriptor was properly initialized, we would need to have is as None

Reply:
I think it seems better to start off with `Device(...).node_desc is None`. I tried changing it to `None` in `Device.__init__` but a few other things broke and left it as-is here.

Really, this draft PR is a MVP and functions enough to demonstrate how the improved joining behavior works in ZHA (which is how I tested it).

 - Devices with missing node descriptors are re-initialized on startup (which queries just the node descriptor) and every time they send data back to zigpy.
 - Re-initialization works normally, so IKEA devices rejoining does nothing but ask the coordinator to fire off a few requests. When they re-join again and announce themselves, it'll work as intended, without waiting for the request retry period to elapse. Perhaps an "initialization delay" would be useful here.

Reply:
Should not re-initialize devices on start-up. Zha hammers the network enough without the extra traffic, better to trigger re-initialize on incoming traffic, just want to make sure that one is not re-entered.